src/pk_core.c

```diff
@@ -648,7 +648,38 @@ DEF(coreExit,
 // String functions.
 // -----------------
 
-// TODO: substring.
+DEF(coreStrSub,
+  "str_sub(str:string, pos:num, len:num) -> string\n"
+  "Returns a substring from a given string supplied. In addition, "
+  "the position and length of the substring are provided when this "
+  "function is called. For example: `str_sub(str, pos, len)`.") {
+  int c = 0;
+  int64_t len, pos;
+
+  String* str = NULL;
+  if (!validateArgString(vm, 1, &str)) return;
+  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
+  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;
+
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Position or length out of range."));
+
+  String* sub = NULL;
+  sub = ALLOCATE_DYNAMIC(vm, String, len + 1, char);
+
+  if (sub == NULL)
+    RET_ERR(newString(vm, "Unable to allocate memory."));
+
+  for (c = 0; c < len; c++) {
+    sub->data[c] = str->data[pos + c - 1];
+  }
+  sub->data[c] = '\0';
```

Comment:
to create a new object, you have to use `new...` function inorder to initialize it and add the object to the VM's objects list. ex: to create a `list` use `newList()`

here you have to use the function  `String* sub = newStringLength(vm, str->data + pos, (uint32_t)len);`